lib/pbench/agent/tool_data_sink.py

```diff
@@ -125,90 +161,101 @@ def __init__(
         self.logger = logger
         self.tool_metadata = tool_metadata
         self.tool_group_dir = self.benchmark_run_dir / f"tools-{self.tool_group}"
-        self.abort_launch = True
 
     def launch(self):
         pass
 
     def terminate(self):
-        if self.abort_launch:
+        if not self.run:
             return 0
 
-        self.run.terminate()
-        self.run.wait()
+        try:
+            self.run.terminate()
+            self.run.wait()
+        except Exception as exc:
+            self.logger.error(
+                "Failed to terminate expected collector process: '%s'", exc
+            )
+            return 0
         return 1
 
 
 class PromCollector(BaseCollector):
+    """Persistent tool data collector for tools compatible with Prometheus"""
+
     def __init__(self, *args, **kwargs):
         super().__init__(*args, **kwargs)
         self.volume = self.tool_group_dir / "prometheus"
 
     def launch(self):
-
-        if self.host_tools_dict:
-            self.abort_launch = False
-        else:
+        if not find_executable("podman"):
+            self.logger.error(
+                "Podman is not installed on this system (required by some registered tools, aborting launch)"
+            )
             return 0
 
-        config = open("prometheus.yml", "w")
-
-        config.write("global:\n  scrape_interval: 1s\n  evaluation_interval: 1s\n\n")
-        # config.write("alerting:\n  alertmanagers:\n  - static_configs:\n    - targets:\n\nrule_files:\n\n")
-        config.write(
-            "scrape_configs:\n  - job_name: 'prometheus'\n    static_configs:\n    - targets: ['localhost:9090']\n\n"
-        )
-
-        for host in self.host_tools_dict:
-            if host.startswith("local"):
-                s = socket.socket(socket.AF_INET, socket.SOCK_DGRAM)
-                s.connect(("8.8.8.8", 80))
-                host_ip = str(s.getsockname()[0])
-                s.close()
-            else:
-                host_ip = host
-
-            for tool in self.host_tools_dict[host]:
-                port = self.tool_metadata.getProperties(tool)["port"]
-                config.write(
-                    "  - job_name: '{}_{}'\n    static_configs:\n    - targets: ['{}:{}']\n\n".format(
-                        host_ip, tool, host_ip, port
-                    )
+        tool_context = []
+        with open("prometheus.yml", "w") as config:
+            for host in self.host_tools_dict:
+                # Used to get ip_addr of unnamed hosts like 'localhost.localdomain'
+                if host.startswith("local"):
+                    s = socket.socket(socket.AF_INET, socket.SOCK_DGRAM)
+                    s.connect(("8.8.8.8", 80))
+                    host_ip = str(s.getsockname()[0])
```

Comment:
Are these three lines (202-204) solely to get the current host IP address?

Reply:
Yeah, other more common methods had failed to work universally, so this was to  ensure uniform results

Reply:
I realize this is in the original code, but I don't think we should keep this anymore.  This is not going to work reliably on hosts with a route that matches the hardcoded `8.8.8.8`.  You'd then be talking through some networking card only to get back to the local host, or worse it gets the wrong IP address where the node-exporter is not listening.

Networking should be properly setup so that prometheus should be able to open a TCP connection to IP resolved from the given name, to then use the HTTP protocol to fetch data from the OpenMetrics end-point.

If a user does not have a reliable name, then they should register that host by IP address and not by name.